src/main/java/com/cloudbees/plugins/credentials/SystemCredentialsProvider.java

```diff
@@ -325,6 +326,11 @@ private synchronized boolean updateCredentials(@NonNull Domain domain, @NonNull
         checkPermission(CredentialsProvider.UPDATE);
         Map<Domain, List<Credentials>> domainCredentialsMap = getDomainCredentialsMap();
         if (domainCredentialsMap.containsKey(domain)) {
+            if (current instanceof IdCredentials && replacement instanceof IdCredentials) {
```

Comment:
Unsure what makes more sense here AND or OR.

Reply:
depends if we should be able to replace an `IdCredential` with a non `IdCredential`
To me adding or removing an ID (replacing an IdCredential with a non IdCredential or vice versa) would sound like a change in the actual Id so `OR` would make sense.

Any migration from a non IdCredential to IdCredential should occur seemlessly in data migration based on `readResolve` generating a unique Id on load, and hence this method should not have to worry about that.

But I think either `AND` or `OR` is acceptable here.

Reply:
Would like to have more opinions here. Maybe @jglick or @daniel-beck ?

Reply:
Tentatively would prefer `||` for the same reasons as James argues for it.